### PR TITLE
Use Bytes for the Haskell backend

### PR DESCRIPTION
**Pull request by ttuegel**
After: #595

### CODE REVIEW

File: data.md [hunk 1]
```diff
@@ -808,17 +808,22 @@ These parsers can interperet hex-encoded strings as `Int`s, `ByteArray`s, and `M
 ```
 
 ```{.k .bytes}
+    syntax String ::= #alignHexString ( String ) [function]
```
Comment (ttuegel) [inline data.md]: `functional`?

File: data.md [hunk 2]
```diff
@@ -517,9 +517,9 @@ The local memory of execution is a byte-array (instead of a word-array).
 -   `#sizeByteArray` calculates the size of a `ByteArray`.
 -   `#padToWidth(N, WS)` and `#padRightToWidth` make sure that a `WordStack` is the correct size.
 
-```{.k .concrete}
+```{.k .bytes}
     syntax ByteArray = Bytes
-    syntax ByteArray ::= ".ByteArray" [function]
+    syntax ByteArray ::= ".ByteArray" [function, functional]
  // --------------------------------------------
```
Comment (ehildenb) [inline data.md]:
```suggestion
 // --------------------------------------------------------
```

File: data.md [hunk 3]
```diff
@@ -807,23 +807,28 @@ These parsers can interperet hex-encoded strings as `Int`s, `ByteArray`s, and `M
...
- // -------------------------------------------------------------
+ // -------------------------------------------------------------------
     rule #parseByteStack(S) => #parseHexBytes(replaceAll(S, "0x", ""))
-    rule #parseHexBytes("") => .ByteArray
-    rule #parseHexBytes(S)  => #parseHexBytes("0" +String S)
-      requires notBool lengthString(S) modInt 2 ==Int 0
-    rule #parseHexBytes(S)  => Int2Bytes(1, #parseHexWord(substrString(S, 0, 2)), BE) +Bytes #parseHexBytes(substrString(S, 2, lengthString(S)))
-      requires lengthString(S) modInt 2 ==Int 0
-       andBool lengthString(S) >Int 0
+
```
Comment (ehildenb) [inline data.md]:
```suggestion
```

File: data.md [hunk 4]
```diff
@@ -807,23 +807,28 @@ These parsers can interperet hex-encoded strings as `Int`s, `ByteArray`s, and `M
...
+    rule #parseHexBytes(S)  => #parseHexBytesAux(#alignHexString(S))
+    rule #parseHexBytesAux("") => .ByteArray
+    rule #parseHexBytesAux(S)  => Int2Bytes(1, String2Base(substrString(S, 0, 2), 16), BE) +Bytes #parseHexBytesAux(substrString(S, 2, lengthString(S)))
+      requires lengthString(S) >=Int 2
```
Comment (ehildenb) [inline data.md]: Please make the `nobytes` version of this code also use the same `#alignHexString` approach, so there is less deviation in the code. `#alignHexString` can be defined in global `k` code block.

**Pull request merged**